test-basic/data-movement-request.js

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright (c) 2020 MarkLogic Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+const testconfig = require('../etc/test-config.js');
+const marklogic = require('../');
+const dbWriter = marklogic.createDatabaseClient(testconfig.restWriterConnection);
+
+const Stream = require('stream');
+const readable = new Stream.Readable({objectMode: true});
+const writable = new Stream.Writable({objectMode: true});
+
+describe('data-movement-requests test', function(){
+
+    it('should write 1000 documents', function(done){
+
+        writable._write = (object, encoding, done) => {
+            done();
+        };
+
+        readable.pipe(writable);
+        createDocumentDescriptors();
+        readable.on('readable', function() {
+            writeDocs(readable.read(), done);
+        });
+        done();
+    });
+});
+
+function createDocumentDescriptors() {
+
+    for(let i=0; i<1000; i++) {
+        const temp = {
+            uri: '/test/dataMovement/requests/'+i+'.json',
+            contentType: 'application/json',
+            content: {[ 'key'+i]: 'value'+i}
+        };
+        readable.push(temp);
+    }
+    readable.push(null);
+}
+
+function writeDocs(record, done) {
+    if(record === null) {
+        return done();
+    }
+
+    dbWriter.documents.write(record).result((output) => writeDocs(readable.read(), done))
+        .catch((err) => console.log(err));
+
+}
+
```

Comment:
Does the current test prototype the approach from the wiki where the application writes to a stream and the API implementation reads from the same stream?

Specifically, I would expect the prototype to do the following:

- write documents to a stream.Writable (possibly a stream.PassThrough)
- read documents from a stream.Readable (possibly the same stream.PassThrough)
- buffer a batch of documents in an array
- submit the batch when the array is full or the stream ends 
- have multiple concurrent writes pending at the same time (similar to the previous prototype)
- call done when the response from the last write returns

Reply:
Hi Erik
The stream "readable" is the stream where data is written to and read from. "createDocumentDescriptors()" writes to this stream (pushes the document descriptors) and then "writeDocs()" reads from this stream and writes to database. So the changes will be - changing the type of "readable" to stream.PassThrough and the batch implementation?

Reply:
Sorry for the lack of clarity.  The purpose of prototyping is to make progress toward the implementation.

The first prototype supported reading batches of documents from an array by independent requests that processed each response by making a new request.

To make progress toward the implementation, it would be great if the second prototype built on the first.  The change would be to read from a stream instead of an array.  Management of concurrency should be similar to the first prototype -- that's the critical aspect of a data movement function.  

How the stream is populated isn't that important.  Ideally, the stream would be populated by writing to a stream.PassThrough, but it's also possible to wrap a stream.Readable implementation around an array.